src/htmlize/patchtml.ml

```diff
@@ -20,10 +19,115 @@ let edit_html ?header ?footer ~head_childs html =
     footer
     "#footer";
   let head = html $ "head" in
-  List.iter (fun (id,child) -> 
-      opt_inserter
-        (iter (fun e -> append_child head e))
-        (Some child)
-        id)
+  List.iter (fun (id,child) ->
+    opt_inserter
+      (iter (fun e -> append_child head e))
+      (Some child)
+      id)
     head_childs;
   to_string html
+
+open Omd
+(* Keeps us from processing url links *)
+let is_url url =
+  (String.length url >= 7 && String.equal (String.sub url 0 7) "http://")
+  || (String.length url >= 8 && String.equal (String.sub url 0 8) "https://")
+
+(* Keeps us from processing anchors *)
+let is_anchor link =
+  String.length link >= 1 && String.equal (String.sub link 0 1) "#"
+
+(* Establish depth of given link relative to the root
+ * Then prepend the corresponding number of .. to the basename
+ * Then append index.html to fix link
+ * *)
+let handle_link link =
+  if link.destination = "" then
+    link
+
+  (* Escape through this branch if the link points to an http(s) address,
+   * or is a link from a Table of Contents
+   * *)
+  else if is_url link.destination || is_anchor link.destination then
+    link
+
+  else
+    (* Turn the link into a list in order to process it *)
+    let li = String.split_on_char '/' link.destination in
+
+    (* Filter out the empty strings resulting from forms like:
+     * path/to/directory/ = [path;to;directory;""]
+     * They're are not counted as far a depth goes.
+     * And they do not change the semantics of the link
+     * *)
+    let llink = List.filter (fun x -> x <> "." && x <> "") li in
+
+    let dep llink =
+      let neg_depth = List.filter (fun x -> x = "..") llink in
+      List.length llink - List.length neg_depth
+    in
+
+    let new_link =
+      let initial_depth = dep llink in
+
+      let rec process link =
+        match link with
+        (* Return reversed link *)
+        | [] -> link
```

Comment:
`assert false` ?

Reply:
Uh well, after checking, it's likely that some cases make the last call to `List.filter` return an empty list, therefore sending into function `process` an empty link

Reply:
Do you have an example of such a case ? Could you add one to the tests ?

Reply:
actually there are some tests currently in `test/patchtmltests.ml`that do so. The first one for example.

Reply:
`[] -> []` would be better